stm32-modules/include/flex-stacker/flex-stacker/tof_sensor_task.hpp

```diff
@@ -184,6 +188,114 @@ class TOFSensorTask {
+        auto sensor = &get_sensor(m.sensor_id);
+        if (m.cancel) {
+            sensor->driver.stop_measurement(m.sensor_id);
+            reset_measurement_state(m.sensor_id);
+            response = messages::StartTOFMeasurementResponse{
+                .responding_to_id = m.id,
+                .sensor_id = m.sensor_id,
+                .cancelled = true,
+                .len = 0,
+            };
+            return send_response(response);
+        }
+
+        if (!sensor->ok || sensor->state == MEASURING) {
+            // TODO: send specific error code when already measuring
+            response = messages::ErrorMessage{
+                .code = errors::ErrorCode::TMC2160_INVALID_ADDRESS};
+            return send_response(response);
+        }
+
+        // Start measurement
+        auto len = sensor->driver.start_measurement(m.sensor_id, m.kind);
+        if (len < 0) {
+            reset_measurement_state(m.sensor_id);
+            // TODO: send specific error code when fail
+            response = messages::ErrorMessage{
+                .code = errors::ErrorCode::TMC2160_INVALID_ADDRESS};
+            return send_response(response);
+        }
+
+        // Success, Set state and send response
+        sensor->state = MEASURING;
+        sensor->message_id = 0;
+        response = messages::StartTOFMeasurementResponse{
+            .responding_to_id = m.id,
+            .sensor_id = m.sensor_id,
+            .kind = m.kind,
+            .len = (uint16_t)len,
+        };
+        send_response(response);
+    }
+
+    auto visit_message(const messages::GetTOFMeasurementMessage& m) -> void {
+        messages::HostCommsMessage response;
+        auto sensor = &get_sensor(m.sensor_id);
+        if (!sensor->ok || sensor->state != MEASURING) {
+            reset_measurement_state(m.sensor_id);
+            // TODO: send specific error code when NOT measuring
+            response = messages::ErrorMessage{
+                .code = errors::ErrorCode::TMC2160_WRITE_ERROR};
+            return send_response(response);
+        }
+
+        // Resend previous chunk
+        if (m.resend && sensor->current_frame.has_value()) {
+            auto c_string = sensor->current_frame.value().c_str();
+            response = messages::GetTOFMeasurementResponse{
+                .responding_to_id = m.id,
+                .sensor_id = m.sensor_id,
+                .id = sensor->message_id,
+                .data = c_string,
+            };
+            return send_response(response);
+        }
+
+        // Get the next histogram chunk
+        auto [ret, data] = sensor->driver.get_histogram_chunk(m.sensor_id);
+        if (ret == HIST_ERROR) {
+            reset_measurement_state(m.sensor_id);
+            // TODO: send specific error code
+            response = messages::ErrorMessage{
+                .code = errors::ErrorCode::TMC2160_INVALID_ADDRESS};
+            return send_response(response);
```

Comment:
Do we need to do something here to signal the host comms task to remove the `get_tof_measurement_cache`, so we can receive and process another measurement request?

Reply:
yeah, good call out.